### PR TITLE
(SIMP-1448) Rework simp config for SIMP 6

### PR DESCRIPTION
**Pull request by lnemsick-simp**
- Output hieradata appropriate for SIMP6 (correct
  names and types).
- Use scenario yaml file as a beginning point for parameter
  values and only prompt for what is missing.
- Set $simp_scenario variable in 'simp' environment's site.pp
- Only output global SIMP hieradata in simp_config_settings.yaml.
- Set SIMP server hieradata in the server's <host>.yaml file,
  instead of SIMP's global hieradata file.
- Copy simp from /usr/share/simp if copy has not yet been done
  for NON-ISO, RPM install.  NOTE:  Does NOT prompt user,
  as simp config operation won't work if those files are not
  in place. SIMP-1561
- Ask user if they want to set simp to the default production
  environment and then create production links to simp in the
  primary and (SIMP-specific) secondary Puppet environments
  directories.
- Rework logic relating to LDAP to distinguish between the
  case in which LDAP server is or is not on the SIMP server.
  - SIMP-647
  - Remove scraping of openlap configuration files to determine
    OS values for LDAP Base DN, Bind DN, RootDN, and server URIs.
    This method of determining LDAP config is unreliable.
- Rework logic related to YUM repos, so that potential problems
  with external repos are highlighted. SIMP-647
  - Detect when external repos are required and setup as
    simp::yum parameters as much as we can
  - Force user to verify, manually, that external repos are
    setup before 'simp bootstrap' can run
- Use current FIPS mode to determine simp_options::fips setting
  instead of giving user the option from 'simp config' to screw
  up their system.
- Bug fixes
  - Don't output text from grub2 configuration, certificate
    creation, and yum repo update SIMP-407, SIMP-409, SIMP-410
  - For ISO/RPM installs (only), handle the problem
    in SIMP-1416. Solution for R10k systems TBD.
  - Don't log intro and summary text when --quiet option is
    specified.
- Minor enhancements for end-user
  - Add log of simp config output, including details of the apply
  - Capture and log error output from spawned processes
    (primarily from ActionItems)
  - Suppress ActionItem non-error console output
  - Backup more of the files modified by simp config and include
    a timestamp in those file names .
  - Use start time of 'simp config' for any backups, so the user
    can correlate all changes 'simp config' made in a run.
  - Clean up text presented to end user
    - Improve output from ActionItems
    - Improve Item descriptions/summaries
    - Make sure PasswordItems tell you which password they are
      querying for, when --quiet mode is selected
  - Write keys in alphabetical order into answers YAML file and
    global hieradata YAML file, so that the user can easily navigate
    these documents.
  - Don't ask about forwarding syslog servers if no syslog servers
    are specified.
- Minor refactoring for maintainability
  - Add IntegerItem to support integer hieradata.
  - Only implement apply in action items.  Allows us to
    use ActionItem.description to describe tersely what is
    happening while it is happening.  (Warm fuzzy for user.)
  - Rename Config::Items for clarity/maintainability
  - Clean up options list for simp config
  - Implement error reporting for Item dependencies.  This makes
    debugging Item tree errors easier.
  - Rework directory layout for Items to facilitate code maintenance.
- Additional tests

SIMP-407 #close
SIMP-409 #close
SIMP-410 #close
SIMP-496 #close
SIMP-647 #close
SIMP-1191 #close
SIMP-1448 #close
SIMP-1540 #close
SIMP-1561 #close

### CODE REVIEW

File: spec/lib/simp/cli/config/items/action/set_production_to_simp_action_spec.rb [hunk 1]
```diff
@@ -0,0 +1,114 @@
+require 'simp/cli/config/items/action/set_production_to_simp_action'
+require_relative '../spec_helper'
+
+describe Simp::Cli::Config::Item::SetProductionToSimpAction do
+  before :each do
+    @ci            = Simp::Cli::Config::Item::SetProductionToSimpAction.new
+#    @ci.silent     = true # comment out this line if you want to see the messages generated
```
Comment (op-ct) [inline spec/lib/simp/cli/config/items/action/set_production_to_simp_action_spec.rb]: Should this be removed, re-enabled, or keyed to an environment variable?

File: spec/lib/simp/cli/config/items/action/check_remote_yum_config_action_spec.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
+require 'simp/cli/config/items/action/check_remote_yum_config_action'
+require_relative '../spec_helper'
+
+describe Simp::Cli::Config::Item::CheckRemoteYumConfigAction do
+  before :each do
+    @ci        = Simp::Cli::Config::Item::CheckRemoteYumConfigAction.new
+#    @ci.silent = true # uncomment out this line to see log message
```
Comment (op-ct) [inline spec/lib/simp/cli/config/items/action/check_remote_yum_config_action_spec.rb]: Should this be removed, re-enabled, or keyed to an environment variable?

File: lib/simp/cli/config/items/action/set_grub_password_action.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      super
+      @key         = 'set_grub_password_action'
+      @description = 'Set GRUB password'
+      @applied_status = :unattempted
+    end
+
+    def apply
+      @applied_status = :failed
+      grub_hash = get_item('grub::password').value
+      if Facter.value('lsbmajdistrelease') > "6" then
```
Comment (op-ct) [inline lib/simp/cli/config/items/action/set_grub_password_action.rb]: In the future, we may need to check that the osfamily is RedHat as well.

File: lib/simp/cli/config/items/action/setup_puppet_autosign_action.rb [hunk 4]
```diff
@@ -1,15 +1,14 @@
...
 class Simp::Cli; end
 module Simp::Cli::Config
-  class Item::PuppetAutosign < ActionItem
+  class Item::SetupPuppetAutosignAction < ActionItem
     attr_accessor :file
     def initialize
       super
       @key         = 'puppet::autosign'
-      @description = %Q{By default, the only host eligible for autosign is the Puppet master; action-only.}
+      @description = 'Setup Puppet autosign'
```
Comment (op-ct) [inline lib/simp/cli/config/items/action/setup_puppet_autosign_action.rb]: "Set up" used as a verb is two words.

File: lib/simp/cli/config/items/action/update_os_yum_repositories_action.rb [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+class Simp::Cli; end
+module Simp::Cli::Config
+  class Item::UpdateOsYumRepositoriesAction < ActionItem
+    class YumRepoError < RuntimeError; end
+
+    attr_accessor :www_yum_dir, :yum_repos_d, :dir
+    def initialize
+      super
+      @key         = 'yum::repositories::update'
+      @description = 'Setup local YUM repositories for SIMP'
```
Comment (op-ct) [inline lib/simp/cli/config/items/action/update_os_yum_repositories_action.rb]: "Set up" used as a verb is two words.

File: spec/lib/simp/cli/commands/config_spec.rb [hunk 6]
```diff
@@ -101,191 +183,356 @@ def normalize(config_lines)
...
+       fmt_begin = "\e[35m\e[1m"
+       fmt_end = "\e[0m"
+       skip_msg = '(Skipping apply[**user is not root**])'
+       expected_lines = [
+         "#{fmt_begin}#{skip_msg}#{fmt_end} Set $simp_scenario in simp environment's site.pp",
+         "#{fmt_begin}#{skip_msg}#{fmt_end} Configure a network interface",
+         "#{fmt_begin}#{skip_msg}#{fmt_end} Set hostname",
+         "#{fmt_begin}#{skip_msg}#{fmt_end} Set GRUB password",
+         "#{fmt_begin}#{skip_msg}#{fmt_end} Set default Puppet environment to 'simp'",
+         "#{fmt_begin}#{skip_msg}#{fmt_end} Setup Puppet autosign",
```
Comment (op-ct) [inline spec/lib/simp/cli/commands/config_spec.rb]: "Set up" used as a verb is two words.

File: lib/simp/cli/config/items/action/add_puppet_hosts_entry_action.rb [hunk 7]
```diff
@@ -1,30 +1,34 @@
...
 module Simp::Cli::Config
-  class Item::PuppetHostsEntry < ActionItem
+  class Item::AddPuppetHostsEntryAction < ActionItem
     attr_accessor :file
 
     def initialize
       super
-      @key         = 'puppet::hosts_entry'
-      @description = %Q{Ensures an entry for the Puppet server in /etc/hosts; action-only.}
+      @key         = "puppet::hosts_entry"
```
Comment (lnemsick-simp) [inline lib/simp/cli/config/items/action/add_puppet_hosts_entry_action.rb]: I have no idea.  Should have been single quotes.  Will fix it back.

File: lib/simp/cli/config/items/action/configure_network_action.rb [hunk 8]
```diff
@@ -19,43 +17,42 @@ def apply
...
-        cmd += %Q@nameservers => #{ format_puppet_array( dns_servers ) }, @
+        cmd += %Q@servers => #{ format_puppet_array( dns_servers ) }, @
         cmd += %Q@search => #{ format_puppet_array( dns_search ) }, @
         cmd += %Q@named_autoconf => false, @
       end
       cmd += %Q@}"@
 # TODO: maybe good ideas
 #   - set $::domain with FACTER_domain=
-#   - set comon::resolv{ named_autofonf => false
+#   - set simplib::resolv{ named_autofonf => false
```
Comment (lnemsick-simp) [inline lib/simp/cli/config/items/action/configure_network_action.rb]: I inherited the TODOs.  If they don't make sense to you, I can take them out.

File: lib/simp/cli/config/items/data/cli_simp_scenario.rb [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+module Simp; end
+class Simp::Cli; end
+module Simp::Cli::Config
+  class Item::CliSimpScenario < Item
+    def initialize
+      super
+      @key         = 'cli::simp::scenario'
+      @description = %Q{SIMP scenario
+
```
Comment (lnemsick-simp) [inline lib/simp/cli/config/items/data/cli_simp_scenario.rb]: I'm not sure I understand the scenario you are envisioning.  Can you please clarify?

File: spec/lib/simp/cli/config/items/action/set_production_to_simp_action_spec.rb (hunk 1 above)
Comment (lnemsick-simp) [inline spec/lib/simp/cli/config/items/action/set_production_to_simp_action_spec.rb]: I should have re-enabled all of those before checking in, although I do like the idea of an environment variable allowing us to selectively debug each class.  That would be a better option than constantly changing the code.

**Comment by nick-markowski** [comment]
the latest iteration worked on centos7 with static interface and essentially defaults for everything else

**Pull request merged**